speechbrain/core.py

```diff
@@ -1050,100 +1068,42 @@ def fit_batch(self, batch):
-            loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    self.scaler.scale(
-                        loss / self.grad_accumulation_factor
-                    ).backward()
-                if should_step:
-                    self.scaler.unscale_(self.optimizer)
-                    self.scaler.step(self.optimizer)
-                    self.scaler.update()
-                    self.zero_grad()
-                    self.optimizer_step += 1
-        else:
-            if self.bfloat16_mix_prec:
-                with torch.autocast(
-                    device_type=torch.device(self.device).type,
-                    dtype=torch.bfloat16,
-                ):
-                    outputs = self.compute_forward(batch, Stage.TRAIN)
-                    loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-            else:
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-                loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
-                if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
-
-        self.on_fit_batch_end(batch, outputs, loss, should_step)
-        return loss.detach().cpu()
+        amp = AutocastConfig.from_name(self.autocast)
+        should_step = (self.step % self.grad_accumulation_factor) == 0
+
+        # We wrap the forward pass in no_sync to avoid
+        # unnecessary synchronization of gradients.
+        # Empirically, we found that this improves performance by a lot (2x speedup).
+        with self.no_sync():
+            with torch.autocast(
+                dtype=amp.dtype, device_type=torch.device(self.device).type,
+            ):
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
 
-    def on_fit_batch_end(self, batch, outputs, loss, should_step):
-        """Called after ``fit_batch()``, meant for calculating and logging metrics.
+        with self.no_sync(not should_step):
+            # The objectives are removed from Autocast to avoid
+            # potential numerical instabilities with CTC loss, etc.
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+            loss = loss / self.grad_accumulation_factor
```

Comment:
Should'nt this be `scaled_loss = self.scaler.scale(loss / self.grad_accumulation_factor)`? Otherwise we return the loss value divided by `self.grad_accumulation_factor` instead of the original value

Reply:
True. Indeed. Thanks for the catch!

Reply:
One question: should we return the scaled_loss or the current loss with the gradscaler scaling ?

Reply:
I think it's OK to return the current loss in every case